examples/threading/threading.c

```diff
@@ -1,33 +1,127 @@
+/*
+*   file:      threading.c
+*   brief:     Creates a thread which sleeps a specified amount of milliseconds before acquiring a mutex, then sleeps a specified amount of milliseconds before releasing it
+*   author:    Guruprashanth Krishnakumar, gukr5411@colorado.edu
+*   date:      09/13/2022
+*   refs:      Ch.7 of Linux System Programming by Robert Love, lecture slides of ECEN 5713 - Advanced Embedded Software Dev.
+*/
+
+/*
+*   HEADER FILES
+*/
 #include "threading.h"
 #include <unistd.h>
 #include <stdlib.h>
 #include <stdio.h>
 
 // Optional: use these functions to add debug or error prints to your application
-#define DEBUG_LOG(msg,...)
-//#define DEBUG_LOG(msg,...) printf("threading: " msg "\n" , ##__VA_ARGS__)
+//#define DEBUG_LOG(msg,...)
+#define DEBUG_LOG(msg,...) printf("threading: " msg "\n" , ##__VA_ARGS__)
 #define ERROR_LOG(msg,...) printf("threading ERROR: " msg "\n" , ##__VA_ARGS__)
+#define NSEC_PER_MSEC           (1000000)
+#define MSEC_PER_SEC            (1000)
 
+/*
+*   FUNCTION DEFINITIONs
+*/
+/*
+*  Wait, obtain mutex, wait, release mutex as described by thread_data structure
+*  args:
+*       thread_params: parameters required for thread execution including the mutex, sleeptimes etc
+*  return:
+        thread_param
+*/
 void* threadfunc(void* thread_param)
 {
 
-    // TODO: wait, obtain mutex, wait, release mutex as described by thread_data structure
-    // hint: use a cast like the one below to obtain thread arguments from your parameter
-    //struct thread_data* thread_func_args = (struct thread_data *) thread_param;
+    //Sleep for specified amount of milliseconds, if the sleep is exited earlier due to any reason (like signals), repeat sleep until full sleep time is achieved
+    struct thread_data* thread_func_args = (struct thread_data *) thread_param;
+    struct timespec sleep_time, remaining_time;
+    sleep_time.tv_sec = thread_func_args->wait_before_obtain.tv_sec;
+    sleep_time.tv_nsec = thread_func_args->wait_before_obtain.tv_nsec;
+    int ret;
+    do
+    {
+        ret=nanosleep(&sleep_time, &remaining_time);
+        if(ret == 0)
+        {
+             break;
+        }
+        sleep_time.tv_sec = remaining_time.tv_sec;
+        sleep_time.tv_nsec = remaining_time.tv_nsec;
+    }
+    while (((remaining_time.tv_sec > 0) || (remaining_time.tv_nsec > 0)));
```

Comment:
@GuruprashanthKrishnakumar rather than duplicate this block at line 65 think about putting in a function you could call.